### DISCUSSION TITLE
How can Supabase say they offer "self-hosting"?

### DISCUSSION THREAD

**Discussion by globemediaofficial** [supabase/supabase#9514] [Questions]
I was super excited to start using Supabase instead of Firebase because it offered "self-hosting". It didn't take me long to realize that "self-hosting" belongs in quotation marks here. To start, the docs were lacking a lot. They say how to start docker and stop there. No information about port forwarding or accessing the API externally.

After figuring it out myself, I now keep discovering key Supabase features that aren't available when self-hosting. There is no storage panel on the studio dashboard, there are no authentication settings, and the list keeps going. This isn't new stuff that they need to create, it's stuff that was intentionally left out, and they have made that clear.

Throughout the past year, people have opened many issues, discovering more and more things that are not available. Members of the Supabase team have responded, always promising to look into it or start working on it immediately, but never following through. The main excuse has been that they have no way to secure things like the studio when it comes to self-hosting but that doesn't matter. If that was really stopping them, they could offer a full version with a warning, along with the currently limited version.

It would be one thing if they said self-hosting was in beta or had a warning in the docs about how many things are missing, but they don't. Instead, it's marketed as something fully functional and easy to set up.

**Comment by Sukikiroi** [comment]
Hello my freind,
about port forwarding or accessing the API externally  is not related to Supabase it's in the Docker Docs.
Also About other requirement they Mention that in this section
![image](https://user-images.githubusercontent.com/47844266/195543430-08a969a2-7ae4-4f3c-8d46-3b5499ea0467.png).

Also you can ask me anything and I am pleased to help you.

**Reply by globemediaofficial** [reply]
So, if “each system can be configured to suit the particular use case,” where are the instructions for that? How am I meant to configure things like Auth and storage? Also, even if accessing the API externally is done through Docker and not Supabase, that is a crucial part of self-hosting. Supabase can (and should) provide instructions for that. Almost nobody self-hosts with the intention of only using it locally.

**Reply by FuadEfendi** [reply]
Instructions are in ".env" file here: https://github.com/supabase/supabase/blob/master/docker/.env.example

And here: https://supabase.com/docs/guides/hosting/docker#get-the-code


I am not sure about the question: Supabase is Docker container, "Supabase Is Docker".

When I start it locally, via CLI `supabase start`, I have this in console, just as a trivial example:
```
         API URL: http://localhost:54321
          DB URL: postgresql://postgres:postgres@localhost:54322/postgres
      Studio URL: http://localhost:54323
    Inbucket URL: http://localhost:54324
```

The rest is outside of Supabase & Docker control: it is TCP/IP/ DNS, and etc.; it is configurable via `.env` file which Docker uses to start services.

Indeed, Supabase shared source code, I can see Studio at port http://localhost:54323/project/default, I can setup firewall between Studio and Public Internet, whitelist few IP addresses, use tunneling, authentication, encryption, VPN, DNS, and so on; but it is outside of Supabase team responsibility, it is environment specific, per use case.

**Comment by FuadEfendi** [comment]
I understand a little more from your question; I have similar concerns too: with "self-hosting", I cannot create organization, project; I cannot hit `http://localhost:3000/account/tokens` to generate tokens; I don't see "Settings" button at bottom of left bar. I just started playing around few hours ago and posted these questions here.

If "self-hosted" has such limitations then it should be documented.

In the future I plan to use Supabase hosting; but I am developer and I need proper development setup; indeed I can contribute to the project.

**Reply by FuadEfendi** [reply]
RE: " There is no storage panel on the studio dashboard"

Here it is: https://github.com/supabase/storage-api
- and check Docker documentation site how to include it. I assume, for self-hosted, you may or may not consider docker, you can install different services on different hosts, and so on. You can for example configure cluster of Postgres with load balancer and fail-over, separate cluster of "Studio", and so on; obviously Supabase Cloud uses similar setup.

And this is what I need right now: 
https://github.com/supabase/supabase/blob/master/DEVELOPERS.md
https://discord.supabase.com/

**Reply by globemediaofficial** [reply]
I am not using AWS and it seems like that's required in that update. I need the storage to be stored in my self-hosted docker container.

**Reply by FuadEfendi** [reply]
Quote: "Integrates with S3 as the storage backend (with more in the pipeline!)"
- they have filesystem-based too
- they also have "generic" interface which you can extend and add custom "storage" (and hopefully share with community)

https://github.com/supabase/storage-api/tree/master/src/backend

**Reply by FuadEfendi** [reply]
And this is how to configure FILE_STORAGE_BACKEND_PATH:
https://github.com/supabase/storage-api/blob/master/.env.sample

**Reply by globemediaofficial** [reply]
What I currently have is the basic Supabase repository with the .env file edited to have my url and passwords/secrets. Where do I go from there? How do I add the storage UI to the dashboard and use the filesystem-based version? This is the kind of stuff that I think is missing in the docs. Good practice for docs is to have them detailed enough that someone with zero experience can get it working.

**Reply by FuadEfendi** [reply]
Yes, it is not documented; but even if it is, best source of documentation is source code... indeed it is just v.0.23.1, expect bugs and changes...

**Reply by globemediaofficial** [reply]
Do I just need to add the variables to the .env file or do I need to update the Supabase directory? How would I do that?

**Reply by FuadEfendi** [reply]
Sorry, I cannot help you. I got familiar with Supabase only today.

**Comment by FuadEfendi** [comment]
See https://github.com/supabase/supabase/discussions/4053
it seems to be a reported bug.

**Reply by globemediaofficial** [reply]
I'm not sure that it's the same issue. I have auth and the database working, they are just very limited when compared to what a Supabase cloud offers.

**Reply by FuadEfendi** [reply]
Can you please be more specific? I noticed some issues which I believe are bugs: for example, Authentication -> Configuration panel is still not available (after I fixed bug in `docker-compose.yml`).

I suggest to start development with "cloud" platform and decide "docker" vs. "free cloud" vs. "paid" vs. "some other offering" later. Use "self hosting" version from "gitlab" only if you are confident with source code; "gitlab" version frequently gets updated and one should be aware of it and be able to contribute fixes and improvements.

**Reply by globemediaofficial** [reply]
What I mean by the self-hosted version being limited is that there is no configuration available in the studio. If I want to enable oauth, I have to figure out what variables to add for each provider. I don’t know why Supabase doesn’t at least provide those variables in the example. Then, for storage, you have to configure it through SQL which is more difficult that being able to see a visual representation of the tables and policies. Also, you can’t have multiple projects. Even when I define the project name and company name variables in the .env file, it’s just the one default project. It all really leads back to the lack of documentation for self-hosting. Technically, everything is there, but it’s not easy to get working.

**Reply by FuadEfendi** [reply]
> If I want to enable oauth, I have to figure out what variables to add for each provider. I don’t know why Supabase doesn’t at least provide those variables in the example.

I believe this is bug in current `master` branch in GitHub, reported at https://github.com/supabase/supabase/discussions/4053


> Then, for storage, you have to configure it through SQL which is more difficult that being able to see a visual representation of the tables and policies. 

I don't think I fully understood your concern. If you run it self-hosted you have 100 GUI tools providing 1000x times richer functionality with database visualizing and design: such as DbVisualizer, IntelliJ, etc.

> Also, you can’t have multiple projects.

It is probably bug, needs to be reported. I am looking into it too.

I assume I need to configure SMTP service; so that users can Register using provided link in Email; upon registration, they will have option to register their organization and project. Right now, I run "docker" as-is, without SMTP, and I connect to "default/default" without authentication. In Chrome debug I can see URL http://localhost:3000/api/profile/permissions?returning=minimal is not available, I need to research it too.

P.S. Of course such open-source products require some extra efforts and technical skills; Supabase needs contributors. 
 It is Apache Licensed, https://github.com/supabase/supabase/blob/master/LICENSE

**Reply by globemediaofficial** [reply]
That’s the issue… these aren’t bugs. I’ve looked through any related discussions on here and some responses from employees say that these configurations were intentionally left out of the self-hosted studio. In their own words, 

> “If it’s in the database, it’s in the studio. If it’s out of the database, it’s not in the studio.”

Since the configurations and storage are not a part of the database, they are not in the studio. I think it’s dumb and they’ve promised for over a year that they’re working on it but I don’t think that’s true.

**Reply by FuadEfendi** [reply]
This is typical and not an issue; de-facto with many similar projects: if you need SLA terms, sign contract; if you have enough technical expertise and want to contribute and even rebrand and resell, the project is provided under Apache 2.0 license; additionally to outdated documentation, you have full access to source code which always goes ahead of documentation; and so on.

It seems Authentication is configurable per-service via YAML config file (and not from database), and documentation lacks details. And I didn't try latest versions from `master` branch yet, I expect they have *super cool* features which are not available for "cloud" yet.

**Comment by berbagimadani** [comment]
Awesome supabase. So what's the conclusion, Does a self hosted supabase have a limit? please explain clearly. Thankyou

**Reply by globemediaofficial** [reply]
Well, in the past few days, storage UI has finally been added, which was something I complained about. Other than that most of the other issues stand. The docs are still very minimal and the only way to get everything working is to look through issues on here and figure it out for yourself. The only other issue I’ve had is that the support is horrible. It would be weeks before you hear back from them (when self-hosting) and your database might be down during that whole time, so it’s just not super reliable. My conclusion is that it’s good IF you want to spend a week or more figuring out everything and if its not gonna be used for something that needs to be super reliable.
